feat: Turnitin API workflow test with openedx XBlock

nandodev-net opened the pull request:
### Turnitin API test xblock
this xblock describes the Turnitin API workflow to generate de similarity report and online viewer.

<p align="center"><img src="https://github.com/eduNEXT/platform-plugin-turnitin/assets/12719909/08fd01bd-b8ec-4133-bb72-9ec8afbd225a" width="600"/></p>

this workflow has 7 steps (handlers) with self explanatory names:

> 1. get_eula_agreement
with a given `eula_version`, request EULA html text to Turnitin API.

> 2. accept_eula_agreement
with a given `eula_version` and `user_id`, accept EULA agreement to Turnitin API.

> 3. 1 . create_turnitin_submission_object
with a user and submission data payload, create a submission object in Turnitin. Response: `submission_id`.

> 3. 2 . upload_turnitin_submission_file
with `submission_id` upload the document assessment.

> 4. get_submission_status
with `submission_id` get the uploaded document status {CREATED - PROCESSING - COMPLETE - ERROR}

> 5. generate_similarity_report
if the submission_status is COMPLETE, with `submission_id` generate the similarity report in Turnitin.

> 6. get_similarity_report_status
with `submission_id` get the similarity report status {PROCESSING - COMPLETE}.

> 7. create_similarity_viewer
if similarity report status is COMPLETE, with `submission_id` get the Viewer URL.

#### Testing

Follow the workflow gif ahead... Secret Key on basecamp

Ian2012 committed:
feat: move work from xblock-test-api

Ian2012 committed:
feat: add initial migrations

nandodev-net committed:
feat: workflow ajax error handle

nandodev-net committed:
fix: File submission response fixed

BryanttV commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,336 @@
+    @XBlock.json_handler
+    def accept_eula_agreement(self, data, suffix=''):
+        user_service = self.runtime.service(self, 'user')
+        user_id = user_service.get_current_user().opt_attrs['edx-platform.user_id']
```
This could be in a function.

BryanttV commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,336 @@
+            current_user = User.objects.get(id=current_user_id)
+            submission = TurnitinSubmission(user = current_user, turnitin_submission_id=turnitin_submission_id)
+            submission.save()
+            print('\n\n')
```
Please remove print statements

BryanttV commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,336 @@
+            #turnitin_submission_id='0a966646-83f9-4ce6-aa47-71e07baf4e30'
+            response = put_upload_submission_file_content(turnitin_submission_id, myfile)
+            return Response(json.dumps(response.json()), content_type='application/json', charset='UTF-8')
+        
```
Please remove trailing whitespace. I recommed using [Trailing Spaces ](https://marketplace.visualstudio.com/items?itemName=shardulm94.trailing-spaces)extension in VSCode

BryanttV commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,336 @@
+        try:
+            last_submission = TurnitinSubmission.objects.filter(user=current_user).latest('created_at')
+        except TurnitinSubmission.DoesNotExist:
+            {'success':False}
```
What is this for?

BryanttV commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,336 @@
+        self.count += 1
+        return {"count": self.count}
+
+
```
Please remove the additional blank lines.

BryanttV commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,336 @@
+    # self.<fieldname>.
+
+    # TO-DO: delete count, and define your own fields.
+    count = Integer(
```
This field is no longer needed, and neither is its handler

BryanttV commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,336 @@
+                "CROSSREF_POSTED_CONTENT"
+            ],
+            "submission_auto_excludes": [
+                "b84b77d1-da0f-4f45-b002-8aec4f4796d6",
```
Where are these UUID's from?

nandodev-net replied:
submission_auto_excludes is an optional field of the endpoint... A list of submission IDs to exclude.

BryanttV commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,336 @@
+            {'success':False}
+        # last_submission = 'de6784c5-471f-4220-aff1-16b6b44dffcf'
+        response = put_generate_similarity_report(last_submission.turnitin_submission_id, payload)
+        if response.status_code == 202:
```
I recommend use `http` module.
```python
from http import HTTPStatus

response.status_code == HTTPStatus.ok 
```

BryanttV commented on platform_plugin_turnitin/turnitin_client/handlers.py:
```diff
@@ -0,0 +1,235 @@
+    TII_API_URL = "https://edunext.tii-sandbox.com"
+    TCA_INTEGRATION_FAMILY = "MySweetLMS"
+    TCA_INTEGRATION_VERSION = "3.2.4"
+    TCA_API_KEY = "[SECRET_KEY]"
```
I think this should be loaded like Django setting

BryanttV reviewed (changes requested):
Please format the code, maybe you could use `black` :thinking:, and don't forget the docstring when everything is ready.

johnvente commented on platform_plugin_turnitin/static/js/src/turnitin.js:
```diff
@@ -0,0 +1,273 @@
+
+  function updateTrafficLightState(semaphoreNumber, state) {
+      console.log(state);
+      const redLight = document.getElementById(`redLight${semaphoreNumber}`);
```
Why don't use jquery here?
```
const redLight = $(`#redLight${semaphoreNumber}`);
```

johnvente commented on platform_plugin_turnitin/static/js/src/turnitin.js:
```diff
@@ -0,0 +1,273 @@
+      const greenLight = document.getElementById(`greenLight${semaphoreNumber}`);
+      const generateReportBtn = document.getElementById(`generateReportBtn${semaphoreNumber}`);
+      const statusText = document.getElementById(`statusText${semaphoreNumber}`);
+      const refreshBtn2 = document.getElementById(`refreshBtn2`);
```
Same in this [comment](https://github.com/eduNEXT/platform-plugin-turnitin/pull/4/files#r1347738955) for these constants

Ian2012 commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,336 @@
+        assert data['hello'] == 'world'
+
+        self.count += 1
+        return {"count": self.count}
```
remove this handler code

nandodev-net committed:
fix: requested changes and docstrings

nandodev-net committed:
fix: set environment variables and format code

Ian2012 commented on platform_plugin_turnitin/models.py:
```diff
@@ -1,3 +1,16 @@
 Database models for platform_plugin_turnitin.
 """
+
+from django.contrib.auth.models import User
```
```suggestion
from django.contrib.auth import get_user_model

User = get_user_model()
```

Ian2012 commented on platform_plugin_turnitin/models.py:
```diff
@@ -1,3 +1,16 @@
+from django.db import models
+
+
+class TurnitinSubmission(models.Model):
```
Add a docstring here

Ian2012 commented on platform_plugin_turnitin/models.py:
```diff
@@ -1,3 +1,16 @@
+    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name="submissions")
+    turnitin_submission_id = models.CharField(max_length=255, blank=True, null=True)
+    turnitin_submission_pdf_id = models.CharField(max_length=255, blank=True, null=True)
+    created_at = models.DateTimeField(auto_now_add=True)
+
```
This model will probably need an index to speed up queries. Can you add a course_id field that's indexed?

nandodev-net replied:
In fact, in the future, I will probably need an item or location identifier (which determines to which ORA it belongs), but that should be defined for the integration iteration with ORA.

mariajgrimaldi commented:
If this is ready for review, please use an adequate name for the PR. Thanks!

mariajgrimaldi commented on platform_plugin_turnitin/models.py:
```diff
@@ -1,3 +1,16 @@
+
+
+class TurnitinSubmission(models.Model):
+    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name="submissions")
```
I think `submissions` is too broad. What about `turnitin_submissions` for the related_name?

mariajgrimaldi commented on platform_plugin_turnitin/settings/common.py:
```diff
@@ -35,3 +35,9 @@ def plugin_settings(settings):  # pylint: disable=unused-argument
     """
+
+    # Configuration variables
+    settings.TII_API_URL = "https://edunext.tii-sandbox.com"
+    settings.TCA_INTEGRATION_FAMILY = "MySweetLMS"
+    settings.TCA_INTEGRATION_VERSION = "3.2.4"
+    settings.TCA_API_KEY = "[SECRET_KEY_HERE]"
```
Let's use the `TURNITIN_` prefix for all the settings

mariajgrimaldi commented on platform_plugin_turnitin/static/html/turnitin.html:
```diff
@@ -0,0 +1,70 @@
+<div class="turnitin_block">
```
This template should be in english, including comments. Remember that this repository is public

mariajgrimaldi commented on platform_plugin_turnitin/static/js/src/turnitin.js:
```diff
@@ -0,0 +1,276 @@
+/* Javascript for TurnitinXBlock. */
```
Same comment about the language here

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,427 @@
+        """
+        user_service = self.runtime.service(self, "user")
+        current_user = user_service.get_current_user()
+        data = {
+            "user_id": current_user.opt_attrs["edx-platform.user_id"],
+            "user_email": current_user.emails[0],
+            "user_name": current_user.full_name.split(),
+        }
+        return data
```
```suggestion
        return {
            "user_id": current_user.opt_attrs["edx-platform.user_id"],
            "user_email": current_user.emails[0],
            "user_name": current_user.full_name.split(),
        }
```

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,427 @@
+            ).latest("created_at")
+        except TurnitinSubmission.DoesNotExist:
+            return {"success": False}
+        # last_submission = 'de6784c5-471f-4220-aff1-16b6b44dffcf'
```
^ I think you missed removing this

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,427 @@
+                    "CROSSREF",
+                    "CROSSREF_POSTED_CONTENT",
+                ],
+                "submission_auto_excludes": [
+                    "b84b77d1-da0f-4f45-b002-8aec4f4796d6",
+                    "b86de142-bc44-4f95-8467-84af12b89217",
```
what's this for?

nandodev-net replied:
@mariajgrimaldi The payloads contain a series of optional parameters; that's a standard request for the purposes of this test.

mariajgrimaldi replied:
I meant, what are those random ids in `submission_auto_excludes`? Do they mean something? Where did we get them?

nandodev-net replied:
no, just sample variables..

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,427 @@
+            dict: The status of the similarity report generation process.
+        """
+
+        payload = {
```
what do these settings mean? should we have a plugin setting called TURNITIN_SIMILARITY_REPORT_SETTINGS for this?

nandodev-net replied:
@mariajgrimaldi Yes... we should define each parameter from the optional settings that Turnitin offers to obtain the result we expect from them.

nandodev-net replied:
@mariajgrimaldi The only issue is that there are nested variables throughout the request structure. So in order to create the settings variable, we have to check the docs to create many variables as need and put them  throughout the request structure. We need more time and owner's review to do this

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,427 @@
+            ).latest("created_at")
+        except TurnitinSubmission.DoesNotExist:
+            return {"success": False}
+        # last_submission = 'de6784c5-471f-4220-aff1-16b6b44dffcf'
```
^ I think you missed removing this

nandodev-net committed:
fix: new related_name TurnitinSubmission model, docstings

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+            "extract_text_only": False,
+            "metadata": {
+                "owners": [
+                    {
+                        "id": user_data["user_id"],
+                        "given_name": user_name[0] if user_name else "no_name",
+                        "family_name": " ".join(user_name[1:])
+                        if len(user_name) > 1
+                        else "no_last_name",
+                        "email": user_data["user_email"],
+                    }
```
I think you can get the `firstname` and `lastname` in `get_user_data` instead of here, so the logic is more encapsulated

mariajgrimaldi commented on platform_plugin_turnitin/turnitin_client/handlers.py:
```diff
@@ -0,0 +1,280 @@
+    pretty_print_response(response)
+
+
+# *---Submissions endpoints ---*
+# Submissions is a Turnithin model that has all relative info to a Assessment send by
+# an student.
```
Maybe it'd be better to create a package called: /handlers so we have modules for each of these interactions:
```
/handlers
  submissions.py
  eula.py
  utils.py
```
It'd read better than separating them with in-line comments. We should use them (in-line comments) when explaining why we're doing something, not to describe it.

nandodev-net replied:
Agree :)

mariajgrimaldi commented on platform_plugin_turnitin/turnitin_client/handlers.py:
```diff
@@ -0,0 +1,280 @@
+    return response
+
+
+def pretty_print_response(response, type_of=""):
+    """debug helper function"""
+    content = response.json()
+    print("\n\n")
+    print(f"------{type_of}------")
+    print("\n\n")
+    print(json.dumps(content, indent=4))
+    print("\n\n")
+    print("------------")
+    print("\n\n")
```
I suggest you use the Python logger instead of prints.

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+                "view_settings": {"save_changes": True},
+            },
+            "author_metadata_override": {
+                "family_name": " ".join(user_name[1:])
+                if len(user_name) > 1
+                else "no_last_name",
+                "given_name": user_name[0] if user_name else "no_name",
```
^ Same comments here. We could do this in the `get_user_data`

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+            ).latest("created_at")
+        except TurnitinSubmission.DoesNotExist:
+            return {"success": False}
+        # last_submission = 'de6784c5-471f-4220-aff1-16b6b44dffcf'
```
^ Same comments here

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+            Response: The response after uploading the file to Turnitin.
+        """
+        turnitin_submission = self.create_turnitin_submission_object()
+        if turnitin_submission.status_code == 201:
```
Can we use the [HTTP Status Code](https://docs.python.org/3/library/http.html) package to check the status instead of 201?

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+        turnitin_submission = self.create_turnitin_submission_object()
+        if turnitin_submission.status_code == 201:
+            turnitin_submission_id = turnitin_submission.json()["id"]
+            current_user_id = self.get_user_data()["user_id"]
+            current_user = User.objects.get(id=current_user_id)
```
We use this pattern a couple of times already. Can we have a method for it? get_django_user or something similar

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+                user=current_user, turnitin_submission_id=turnitin_submission_id
+            )
+            submission.save()
+            myfile = data.params["myfile"].file
```
can we change the `myfile` naming?

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+            )
+            submission.save()
+            myfile = data.params["myfile"].file
+            # turnitin_submission_id='0a966646-83f9-4ce6-aa47-71e07baf4e30'
```
^ same comment here, I think we should remove the like altogether

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+                turnitin_submission_id, myfile
+            )
+            return Response(
+                json.dumps(response.json()),
```
if you do response.txt don't you already get the json format? Not sure though, you might need to check

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+                turnitin_submission_id, myfile
+            )
+            return Response(
+                json.dumps(response.json()),
```
I think you can use `json_body` instead

mariajgrimaldi replied:
And I don' think you need to specify content_type and charset

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+                charset="UTF-8",
+            )
+        return Response(
+            json.dumps(turnitin_submission.json()),
```
Same comment about `json_body`

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+            return {"success": False}
+        # last_submission = 'de6784c5-471f-4220-aff1-16b6b44dffcf'
+        response = get_submission_info(last_submission.turnitin_submission_id)
+        return response.json()
```
Can we explicitly return a dictionary? so we know what we're returning. Right now I'm not sure what response.json() returns

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -0,0 +1,426 @@
+            return {"success": False}
+        # last_submission = 'de6784c5-471f-4220-aff1-16b6b44dffcf'
+        response = get_similarity_report_info(last_submission.turnitin_submission_id)
+        return response.json()
```
Same comment here about the `response.json()`

mariajgrimaldi commented on platform_plugin_turnitin/turnitin_client/handlers.py:
```diff
@@ -0,0 +1,280 @@
+    - ValueError: If an unsupported request method is provided.
+    """
+
+    # Configuration variables
```
We shouldn't use in-line comments for descriptions but for explaining why we're doing something, specifically when the code isn't explicit enough.

mariajgrimaldi commented on platform_plugin_turnitin/turnitin_client/handlers.py:
```diff
@@ -0,0 +1,280 @@
+    """
+
+    # Configuration variables
+    TII_API_URL = getattr(settings, "TURNITIN_TII_API_URL", None)
+    TCA_INTEGRATION_FAMILY = getattr(settings, "TURNITIN_TCA_INTEGRATION_FAMILY", None)
+    TCA_INTEGRATION_VERSION = getattr(settings, "TURNITIN_TCA_INTEGRATION_VERSION", None)
+    TCA_API_KEY = getattr(settings, "TURNITIN_TCA_API_KEY", None)
```
Although these are constants, we don't usually use uppercase unless we're declaring them on top of the file (module level): https://peps.python.org/pep-0008/#constants

mariajgrimaldi commented on platform_plugin_turnitin/turnitin_client/handlers.py:
```diff
@@ -0,0 +1,280 @@
+    TCA_INTEGRATION_VERSION = getattr(settings, "TURNITIN_TCA_INTEGRATION_VERSION", None)
+    TCA_API_KEY = getattr(settings, "TURNITIN_TCA_API_KEY", None)
+
+    # Headers configuration
```
Same comment here (and in the rest) about in-line comments

mariajgrimaldi commented on platform_plugin_turnitin/turnitin_client/handlers.py:
```diff
@@ -0,0 +1,280 @@
+        return response
+
+    # Mapping of methods to corresponding functions in the requests library
+    method_map = {
+        "get": requests.get,
+        "post": requests.post,
+        "put": requests.put,
+        "delete": requests.delete,
+        "patch": requests.patch,
+    }
```
Instead of having a method_map, you could pass directly the method:
`turnitin_api_handler(requests.get, ...)`

nandodev-net replied:
Hmm... I believe it would be better to create the method_map within a separate function. This way, we can modify the method conditionals based on a list of strings, rather than a list of methods. This approach:
 `if request_method.lower() in ["post", "put", "patch"]`  
is more elegant than this:
 `if request_method in [requests.patch, request.post,....] `
and certainly cleaner than a series of:
`if request_method == requests.post: 
    ...
elif ... : 
    ...
else:
    ...
`

mariajgrimaldi commented on platform_plugin_turnitin/turnitin_client/handlers.py:
```diff
@@ -0,0 +1,280 @@
+    if not method_func:
+        raise ValueError(f"Unsupported request method: {request_method}")
+
+    # Calling the appropriate method
+    if request_method.lower() in ["post", "put", "patch"]:
+        response = method_func(
+            f"{TII_API_URL}/api/v1/{url_prefix}", headers=headers, json=data
+        )
+    else:
+        response = method_func(
+            f"{TII_API_URL}/api/v1/{url_prefix}", headers=headers, params=data
+        )
```
```
url = f"{TII_API_URL}/api/v1/{url_prefix}"
request_load = {
 "headers": headers
}
if params:
   request_load["params"] = params
if data:
   request_load["data"] = data
response = method_func(url, **request_load) 
```
When using get, the caller send params as an argument. When using the rest the caller sends data. What do you think?

nandodev-net replied:
I like it :) 
I propose this

    args = {
        "headers": headers,
        "json" if request_method.lower() in ["post", "put", "patch"] else "params": data
    }

    response = method_func(f"{TII_API_URL}/api/v1/{url_prefix}", **args)

mariajgrimaldi commented on platform_plugin_turnitin/turnitin_client/handlers.py:
```diff
@@ -0,0 +1,280 @@
+            f"{TII_API_URL}/api/v1/{url_prefix}", headers=headers, params=data
+        )
+
+    return response
```
should this be `response.json()`?

nandodev-net replied:
there's a couple of endpoints in a different format, that's why I use .json() in some @XBlock.json_handler methods, and not in the turnitin_api_handler method

nandodev-net committed:
fix: handlers refactor and format

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -96,10 +97,14 @@ def get_user_data(self):
             "user_id": current_user.opt_attrs["edx-platform.user_id"],
             "user_email": current_user.emails[0],
-            "user_name": current_user.full_name.split(),
+            "user_name": user_full_name[0] if user_full_name else "no_name",
+            "user_last_name": " ".join(user_full_name[1:])
+            if len(user_full_name) > 1
+            else "no_last_name",
```
I don't think you need to prefix all with `user` since get_user_data is pretty explicit.
```suggestion
            "name": full_name[0] if full_name else "",
            "lastname": full_name[1] if full_name[1:] else "",
```

mariajgrimaldi commented on platform_plugin_turnitin/turnitin.py:
```diff
@@ -96,10 +97,14 @@ def get_user_data(self):
         """
         user_service = self.runtime.service(self, "user")
         current_user = user_service.get_current_user()
+        user_full_name = current_user.full_name.split()
```
```suggestion
        full_name = current_user.full_name.split(" ", 1)
```

nandodev-net committed:
fix: code upgrades
- turnitin:get_user_data() : delete 'user_' prefix.
- turnitin:get_django_user(): method created.
- api_hander: get_request_method_func(): method created.
- api_handler: turnitin_api_handler(): changes abour args.

Ian2012 committed:
chore: reformat for quality checks

Ian2012 committed:
docs: fix indentation issues

Ian2012 committed:
chore: reformat for quality checks again

mariajgrimaldi reviewed (approved):
What we'll use for the integration looks good! Thank you

Ian2012 commented:
@mariajgrimaldi please let me know if we can merge this one
